src/hastings_vacuum.erl

```diff
@@ -93,38 +111,62 @@ code_change(_OldVsn, St, _Extra) ->
 
 
 handle_db_event(DbName, created, _St) ->
-    gen_server:cast(?MODULE, {cleanup, DbName}),
+    gen_server:cast(?MODULE, {cleanup, DbName, []}),
     {ok, nil};
 handle_db_event(DbName, deleted, _St) ->
-    gen_server:cast(?MODULE, {cleanup, DbName}),
+    gen_server:cast(?MODULE, {cleanup, DbName, [{context, delete}]}),
     {ok, nil};
 handle_db_event(_DbName, _Event, _St) ->
     {ok, nil}.
 
 
-maybe_start_cleaner(#st{cleaner=undefined}=St) ->
-    case queue:is_empty(St#st.queue) of
+maybe_start_selective_cleaner(#st{selective_cleaner=undefined}=St) ->
+    case queue:is_empty(St#st.qSelectiveClean) of
         false ->
-            start_cleaner(St);
+            start_selective_cleaner(St);
         true ->
             {noreply, St}
     end;
 
-maybe_start_cleaner(St) ->
+maybe_start_selective_cleaner(St) ->
     {noreply, St}.
 
 
-start_cleaner(St) ->
-    {{value, DbName}, NewQ} = queue:out(St#st.queue),
-    {Pid, _} = spawn_monitor(?MODULE, clean_db, [DbName]),
-    {noreply, St#st{queue = NewQ, cleaner = Pid}}.
+maybe_start_full_cleaner(#st{full_cleaner=undefined}=St) ->
+    case queue:is_empty(St#st.qFullClean) of
+        false ->
+            start_full_cleaner(St);
+        true ->
+            {noreply, St}
+    end;
 
+maybe_start_full_cleaner(St) ->
+    {noreply, St}.
+
+
+start_selective_cleaner(St) ->
+    {{value, DbName}, NewQ} = queue:out(St#st.qSelectiveClean),
+    {Pid, _} = spawn_monitor(?MODULE, clean_db, [DbName, []]),
+    {noreply, St#st{qSelectiveClean = NewQ, selective_cleaner = Pid}}.
 
-clean_db(DbName) ->
-    {ok, JsonDDocs} = get_ddocs(DbName),
-    DDocs = [couch_doc:from_json_obj(DD) || DD <- JsonDDocs],
-    ActiveSigs = lists:usort(lists:flatmap(fun active_sigs/1, DDocs)),
-    cleanup(DbName, ActiveSigs).
+
+start_full_cleaner(St) ->
+    {{value, DbName}, NewQ} = queue:out(St#st.qFullClean),
+    {Pid, _} = spawn_monitor(?MODULE, clean_db, [DbName, [{context, delete}]]),
+    {noreply, St#st{qFullClean = NewQ, full_cleaner = Pid}}.
+
+
+clean_db(DbName, Options) ->
+    Context = couch_util:get_value(context, Options, compaction),
+    case Context =:= delete of
+        false ->
+            {ok, JsonDDocs} = get_ddocs(DbName),
+            DDocs = [couch_doc:from_json_obj(DD) || DD <- JsonDDocs],
+            ActiveSigs = lists:usort(lists:flatmap(fun active_sigs/1, DDocs));
+        true ->
+            ActiveSigs = []
+    end,
+    cleanup(DbName, ActiveSigs, Options).
```

Comment:
Rather than have cleanup do the switching logic between the kind of deletions to perform it'd probably be better to have that logic here. Something like:

```erlang
Context = couch_util:get_value(context, Options, compaction),
DoRecovery = config:get_boolean("couchdb", "enable_database_recovery", false),
case {Context, DoRecovery} of
    {delete, true} ->
        move_all_indexes(DbName);
    {delete, false} ->
        delete_all_indexes(DbName);
    _ ->
        delete_inactive_indexes(DbName, get_active_sigs())
end.
```